### ISSUE TITLE
[Feature Request] list all checkpoints

### ISSUE THREAD

**Issue by albers** [docker/docker#30443]
`docker checkpoint ls` lists the checkpoints _for a given container_:

```bash
$ docker checkpoint ls --help | grep Usage
Usage:  docker checkpoint ls [OPTIONS] CONTAINER
```
I have a usecase where I need a full list of _all checkpoints_: bash completion for `docker start --checkpoint`. At the point where the user specifies `--checkpoint`, the container is yet unknown.
Without this feature, I have to iterate over all containers and call `docker checkpoint ls` on each of them, which would result in lots of API calls.

**Comment by thaJeztah** [comment]
Yes, I noticed this is missing as well; the problem, is that in the current implementation, checkpoints are stored per-container, so there is no "global" list of checkpoints. For example, I can create two checkpoints with the same name, but on different containers.

We should look into the design of the checkpoint feature as a whole (there are other issues as well), which is a reason for putting it in experimental for now.

**Comment by thaJeztah** [comment]
ping @boucher @xemul @tiborvass ptal wondering if we should use a different storage layout for checkpoints, e.g. 

    /var/lib/docker/checkpoints/<container-id>/<checkpoint-name>

Or if we want names to be globally unique (like we have for other objects, such as `container`, `volume`;

    /var/lib/docker/checkpoints/<checkpoint-name>/<container-id>/

Last one may be more difficult, as it's more difficult to verify a checkpoint belongs to a certain container.

Also, ideally, checkpoints should have an _id_ (to bring parity with other object types)

**Comment by boucher** [comment]
I think, ultimately, that yes it makes sense for checkpoints to be more of a global resource than a container specific one. The reason we hesitated for now I think is that there's not really a mechanism in place for figuring out which checkpoints could possibly be applied to which containers.

**Comment by tswift242** [comment]
I think this gets back to the conversation in https://github.com/docker/docker/issues/26451. I agree that checkpoints should be a global resource, rather than per-container.

**Comment by thaJeztah** [comment]
Checkpoints should still be per-container, as they are most likely tied to the filesystem as well. Being able to have a global overview of all checkpoints for all containers would make sense though.

**Comment by tswift242** [comment]
They are tied to the fileystem. However, to me, that means they're tied to the docker image from which the checkpoint was created, not to the container from which the checkpoint was created. I think that's an important thing to distinguish, e.g. for cloning use cases.

**Comment by tswift242** [comment]
The direction I propose this go in is associating each checkpoint to a docker image at checkpoint creation time. For the simple case where the filesystem isn't modified between container creation and checkpoint creation, the associated docker image is simply the image of the container that was checkpointed.

The case where the filesystem is modified in between container and checkpoint creation necessitates a new docker image be created which is consistent with the checkpoint, and that's a little different. For that case, I propose a new ```--commit <image_name>``` flag for the checkpoint create command which will create a new docker image from the current state of the filesystem (e.g. docker commit) and then create the checkpoint. In this case, the checkpoint is associated with the newly created docker image. Making the commit and checkpoint creation a single atomic operation has other advantages as well. Currently, if you do a ```docker commit``` and then a ```docker checkpoint create```, there's a small period of time where the process in the container becomes unpaused, which means that the filesystem could be altered in between the commit and checkpoint creation (e.g. application writes to a log file). If this happens, the restore will fail as the criu image and docker image will be inconsistent. Combining the two actions into a single atomic operation where the container process is paused throughout is the only way to guarantee this won't happen.

Once the checkpoint is associated to a docker image, the validation (on ```docker start --checkpoint <checkpoint_name>```) becomes just checking that the image associated to the checkpoint matches the container's image. Then we know the filesystems are compatible. Perhaps there's other validation that needs to happen too, but that's at least a start.

WDYT @boucher @thaJeztah

**Comment by thaJeztah** [comment]
@tswift242 I think that change is orthogonal to this feature request; could you reopen as a new issue?

**Comment by xemul** [comment]
I agree with @tswift242 , checkpoint of container is heavily tied with the FS snapshot. One thing to add here is that CRIU images can also be stacked (incremental dumps or pre-dumps/dump), so the c/r image effectively has two "parents" -- FS snapshot and CRIU parent image set.

As far as live migration is concerned, I don't think that we need to actually _share_ the docker image, just pulling one and syncing missing (ephemeral) layers will be enough. But this anyway means, that we need to couple checkpoint image with FS image.

**Comment by tswift242** [comment]
@thaJeztah Created https://github.com/docker/docker/issues/30507
